lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,358 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Utils.h"
+
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace glow;
+using llvm::isa;
+
+/// A graph with BFS oder.
+struct BFSLevel {
+  /// A list of <level, nodelist> with BFS order.
+  std::vector<std::pair<int, std::vector<Node *>>> levels;
+  /// A set of visited nodes.
+  std::unordered_set<const Node *> visited;
+};
+
+Partitioner::Partitioner(Module *parent, std::vector<DeviceInfo> &devices)
+    : deviceInfo_(devices) {
+  module_ = parent;
+  // Check if all the functions are in the same structure.
+  if (!isSameFunctionFamily()) {
+    assert("not in the same function family");
+  }
+  memSize_ = module_->getConstantsSize();
+}
+
+/// Get the representative function (the one with the largest input) and update
+/// the memSize_.
+void Partitioner::selectRepFunc() {
```

Comment:
this should be static and have module as input, this way you can test it properly.